File: src/ee/ads/MultiAdView.cpp
//
//  MultiBannerAd.cpp
//  ee_x
//
//  Created by Zinge on 10/18/17.
//
//

#include "ee/ads/MultiAdView.hpp"

#include <ee/core/NoAwait.hpp>
#include <ee/core/ObserverHandle.hpp>
#include <ee/core/SpinLock.hpp>
#include <ee/core/Task.hpp>

namespace ee {
namespace ads {
using Self = MultiAdView;

Self::MultiAdView() {
    anchor_ = std::make_pair(0, 0);
    position_ = std::make_pair(0, 0);
    visible_ = false;
    handle_ = std::make_unique<ObserverHandle>();
    locker_ = std::make_unique<SpinLock>();
}

Self::~MultiAdView() = default;

Self& Self::addItem(const std::shared_ptr<IAdView>& item) {
    items_.push_back(item);
    item->setVisible(visible_);
    (*handle_) //
        .bind(*item)
        .addObserver({
            .onLoaded =
                [this, item] {
                    std::unique_lock<SpinLock> lk(*locker_);
                    loadedItems_.insert(item);
                    lk.unlock();

                    // Propagation.
                    dispatchEvent([](auto&& observer) {
                        if (observer.onLoaded) {
                            observer.onLoaded();
                        }
                    });
                },
            .onClicked =
                [this] {
                    // Propagation.
                    dispatchEvent([](auto&& observer) {
                        if (observer.onClicked) {
                            observer.onClicked();
                        }
                    });
                },
        });
    return *this;
}

void Self::destroy() {
    for (auto&& item : items_) {
        item->destroy();
    }
    items_.clear();
    handle_->clear();
}

bool Self::isLoaded() const {
    for (auto&& item : items_) {
        if (item->isLoaded()) {
            return true;
        }
    }
    return false;
}

Task<bool> Self::load() {
    bool result = false;
    for (auto&& item : items_) {
        if (item == activeItem_ && visible_) {
            // Ignore displaying item.
            continue;
        }
        std::unique_lock<SpinLock> lk(*locker_);
        if (loadedItems_.count(item) != 0) {
            // Already loaded and not displayed.
            continue;
        }
        lk.unlock();
        if (co_await item->load()) {
            result = true;
        }
    }
    co_return result;
}

std::pair<float, float> Self::getAnchor() const {
    return anchor_;
}

void Self::setAnchor(float x, float y) {
    anchor_ = std::make_pair(x, y);
    for (auto&& item : items_) {
        item->setAnchor(x, y);
    }
}

std::pair<float, float> Self::getPosition() const {
    return position_;
}

void Self::setPosition(float x, float y) {
    position_ = std::make_pair(x, y);
    for (auto&& item : items_) {
        item->setPosition(x, y);
    }
}

std::pair<float, float> Self::getSize() const {
    // Combined size of all ad views.
    int width = 0;
    int height = 0;
    for (auto&& item : items_) {
        int itemWidth, itemHeight;
        std::tie(itemWidth, itemHeight) = item->getSize();
        width = std::max(width, itemWidth);
        height = std::max(height, itemHeight);
    }
    return std::pair(width, height);
}

void Self::setSize(float width, float height) {
    for (auto&& item : items_) {
        item->setSize(width, height);
    }
}

bool Self::isVisible() const {
    return visible_;
}

void Self::setVisible(bool visible) {
    visible_ = visible;
    for (auto&& item : items_) {
        item->setVisible(false);
    }
    if (visible) {
        std::scoped_lock<SpinLock> lk(*locker_);
        if (loadedItems_.empty()) {
            for (auto&& item : items_) {
                if (item != activeItem_ && item->isLoaded()) {
                    activeItem_ = item;
                    break;
                }
            }
        } else {
            // Prefer to displaying loaded ad.
            for (auto&& item : items_) {
                if (loadedItems_.count(item) != 0) {
                    loadedItems_.erase(item);
                    activeItem_ = item;
                    break;
                }
            }
        }
        if (activeItem_) {
            activeItem_->setVisible(true);
        }
    } else {
        if (activeItem_) {
            // Reload the currently active ad.
            noAwait(activeItem_->load());
        }
    }
}
} // namespace ads
} // namespace ee

File: src/unity/Runtime/Ads/MultiAdView.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace EE {
    public class MultiAdView : ObserverManager<AdObserver>, IAdView {
        private (float, float) _anchor;
        private (float, float) _position;
        private bool _visible;
        private readonly List<IAdView> _items;
        private readonly ObserverHandle _handle;
        private IAdView _activeItem;
        private readonly HashSet<IAdView> _loadedItems;

        public MultiAdView() {
            _items = new List<IAdView>();
            _handle = new ObserverHandle();
            _loadedItems = new HashSet<IAdView>();
        }

        public MultiAdView AddItem(IAdView item) {
            _items.Add(item);
            item.IsVisible = _visible;
            _handle.Bind(item).AddObserver(new AdObserver {
                OnLoaded = () => {
                    _loadedItems.Add(item);
                    DispatchEvent(observer => observer.OnLoaded?.Invoke());
                },
                OnClicked = () => DispatchEvent(observer => observer.OnClicked?.Invoke())
            });
            return this;
        }

        public void Destroy() {
            foreach (var item in _items) {
                item.Destroy();
            }
            _items.Clear();
            _handle.Clear();
            _activeItem = null;
            _loadedItems.Clear();
        }

        public bool IsLoaded => _items.Any(item => item.IsLoaded);

        public async Task<bool> Load() {
            var result = false;
            foreach (var item in _items) {
                if (item == _activeItem && _visible) {
                    // Ignore displaying item.
                    continue;
                }
                if (_loadedItems.Contains(item)) {
                    // Already loaded and not displayed.
                    continue;
                }
                if (await item.Load()) {
                    result = true;
                }
            }
            return result;
        }

        public (float, float) Anchor {
            get => _anchor;
            set {
                _anchor = value;
                foreach (var item in _items) {
                    item.Anchor = value;
                }
            }
        }

        public (float, float) Position {
            get => _position;
            set {
                _position = value;
                foreach (var item in _items) {
                    item.Position = value;
                }
            }
        }

        public (float, float) Size {
            get {
                // Combined size of all ad views.
                var width = 0.0f;
                var height = 0.0f;
                foreach (var item in _items) {
                    var (itemWidth, itemHeight) = item.Size;
                    width = Math.Max(width, itemWidth);
                    height = Math.Max(height, itemHeight);
                }
                return (width, height);
            }
            set {
                foreach (var item in _items) {
                    item.Size = value;
                }
            }
        }

        public bool IsVisible {
            get => _visible;
            set {
                _visible = value;
                foreach (var item in _items) {
                    item.IsVisible = false;
                }
                if (_visible) {
                    if (_loadedItems.Count == 0) {
                        foreach (var item in _items) {
                            if (item != _activeItem && item.IsLoaded) {
                                _activeItem = item;
                                break;
                            }
                        }
                    } else {
                        // Prefer to displaying loaded ad.
                        foreach (var item in _items) {
                            if (_loadedItems.Contains(item)) {
                                _loadedItems.Remove(item);
                                _activeItem = item;
                                break;
                            }
                        }
                    }
                    if (_activeItem != null) {
                        _activeItem.IsVisible = true;
                    }
                } else {
                    if (_activeItem != null) {
                        // Reload the currently active ad.
                        _ = _activeItem.Load();
                    }
                }
            }
        }
    }
}